benchmark/single-source/Flatten.swift

```diff
@@ -0,0 +1,962 @@
+
+    let f = input.flatMap { $0 }
+
+    let cl = f.reduce(into: CountLast(), CountLast.collect)
+    CheckResults((cl.count, cl.last!) == (flatSize, lastElement))
+  }
+}
+
+@inline(never)
+public func run_FlattenAnySeqSeqArray4_joined(_ N: Int) {
+  for _ in 1...N {
+    let input = AnySequence(sequenceArray4)
+
+    let f = input.joined()
+
+    let cl = f.reduce(into: CountLast(), CountLast.collect)
+    CheckResults((cl.count, cl.last!) == (flatSize, lastElement))
+  }
+}
+
+@inline(never)
+public func run_FlattenAnySeqSeqArray4_forinElem(_ N: Int) {
+  for _ in 1...N {
+    let input = AnySequence(sequenceArray4)
+
+    var f: [Int] = []
+
+    for array in input {
+      for element in array {
+        f.append(element)
+      }
+    }
+
+    let cl = f.reduce(into: CountLast(), CountLast.collect)
+    CheckResults((cl.count, cl.last!) == (flatSize, lastElement))
+  }
+}
+
+@inline(never)
+public func run_FlattenAnySeqSeqArray4_forinBatch(_ N: Int) {
+  for _ in 1...N {
+    let input = AnySequence(sequenceArray4)
+
+    var f: [Int] = []
+
+    for array in input {
+      f += array
+    }
+
+    let cl = f.reduce(into: CountLast(), CountLast.collect)
+    CheckResults((cl.count, cl.last!) == (flatSize, lastElement))
+  }
+}
+
+@inline(never)
+public func run_FlattenArrayColorVal_flatMap(_ N: Int) {
+  for _ in 1...N {
+    let input = arrayColorVals
+
+    let f = input.flatMap { [$0.a, $0.r, $0.g, $0.b] }
+
+    CheckResults((f.count, f.last!) == (flatSize, lastBlue))
+  }
+}
+
+@inline(never)
+public func run_FlattenArrayColorVal_map_joined(_ N: Int) {
+  for _ in 1...N {
+    let input = arrayColorVals
+
+    let f = input.map({ [$0.a, $0.r, $0.g, $0.b] }).joined()
+
+    CheckResults((f.count, f.last!) == (flatSize, lastBlue))
+  }
+}
+
+@inline(never)
+public func run_FlattenArrayColorVal_lazy_map_joined(_ N: Int) {
+  for _ in 1...N {
+    let input = arrayColorVals
+
+    let f = input.lazy.map({ [$0.a, $0.r, $0.g, $0.b] }).joined()
+
+    let cl = f.reduce(into: CountLast(), CountLast.collect)
+    CheckResults((cl.count, cl.last!) == (flatSize, lastBlue))
+  }
+}
+
+@inline(never)
+public func run_FlattenArrayColorVal_forinNaive(_ N: Int) {
+  for _ in 1...N {
+    let input = arrayColorVals
+
+    var f: [UInt8] = []
+
+    for c in input {
+      f.append(c.a)
+      f.append(c.r)
+      f.append(c.g)
+      f.append(c.b)
+    }
+
+    CheckResults((f.count, f.last!) == (flatSize, lastBlue))
+  }
+}
+
+@inline(never)
+public func run_FlattenArrayColorValUnsafe(_ N: Int) {
+  for _ in 1...N {
+    let input = arrayColorVals
+
+    let f = input.withUnsafeBufferPointer {
+      $0.baseAddress!.withMemoryRebound(to: UInt32.self, capacity: input.count) {
+        UnsafeBufferPointer(start:$0, count: input.count).flatMap {
+          (rawColor: UInt32) -> [UInt8] in
+          let rgba = rawColor.bigEndian
+          let argb = (rgba << 24) | (rgba >> 8) // rotate
+          return stride(from:24, through:0, by: -8).map {
```

Comment:
This would be so much easier to read of the shift operations are just written out explicitly.

Reply:
I don’t know what you mean. This is as explicit as I could do it. I’m asking for help with the `Color.Unsafe`, because I don’t like any of the 3 I did. None feels canonical.

Reply:
I mean:
 `[ UInt8(truncatingIfNeeded: argb >> 24,
     UInt8(truncatingIfNeeded: argb >> 16,
 ...]`

Why are there two versions of `run_FlattenArrayColorValUnsafe2`?

If you insist on loading `UInt32` then loading `rgba.bigEndian` should rectify that in a platform independent way right?

Reply:
I don’t insist on anything here. This version performs orders of magnitude slower than naive for-in loop. It’s here as a witness of how I’ve failed to apply the pattern you’ve taught me with `Unsafe.Tuple4` to this new byte swizzling scenario. As I said, this is my first Unsafe rodeo…

Reply:
That’s basically what `Flatten.Array.Color.Val.flatMap` does. I wouldn’t expect this to gain any performance advantage from going through Unsafe. So no. My goal with Unsafe variant is to demonstrate the fastest possible way to get to the `[UInt8]` by skipping as much overhead as possible by leveraging knowledge about `ColorVal`’s memory layout etc. Unless it beats imperative for-in by a good margin, there’s no point in including it. It should establish a goal for best-case optimized `flatMap`.

Reply:
This is why I asked you to explain why you're adding each test case. I have no idea what overhead you think you're removing in each case.

Do you have a faster implementation than the one I posted?

Reply:
> If you insist on loading UInt32 then loading rgba.bigEndian should rectify that in a platform independent way right?

I'm not sure I'm using the right assumptions here. As I said, I had to use the `.bigEndian` conversion here, to make it work correctly (by trial and error). But I doubt this achieves true platform independence… You said in https://github.com/apple/swift/pull/20116#issuecomment-437258412 regarding the `Tuple4.Unsafe`:
> Note: Swift does not provide any formal layout rule that says an array of homogeneous tuples is laid out the same as an array of those tuple elements. But with builtin integer types I feel that you're safe making this assumption.

So I think the platform dependent and totally un-guaranteed assumption I'm making here is about the struct endianness or the order of the 4 `Int`s packed in the `struct ColorVal`. Which makes me think that this particular benchmark should be enclosed in `#if #else` block to be x86 specific for now… or is there some layout guarantee I can base this on?

> Do you have a faster implementation than the one I posted?

Unsafe2 and 3 are both faster than yours (Unsafe4-atrick), which is only slightly faster than `flatMap` and worse than the naive imperative approach.

### Flatten.Array.Color.Val
| Variant | Min | Q1 | Med | Q3 | Max |
|---|---:|---:|---:|---:|---:|
| Unsafe | 9622 | 9843 | 10056 | 10396 | 13497 |
| Unsafe2 | 290 | 290 | 291 | 291 | 296 |
| Unsafe3 | 352 | 354 | 355 | 365 | 445 |
| Unsafe4-atrick | 517 | 518 | 526 | 534 | 603 |
| flatMap | 551 | 552 | 559 | 565 | 580 |
| for-in.Naive | 465 | 465 | 477 | 893 | 1016 |
| lazy.flatMap | 7098 | 7180 | 7273 | 7440 | 7704 |
| lazy.map.joined | 7073 | 7207 | 7334 | 7507 | 8096 |
| map.joined | 9768 | 9956 | 10237 | 10511 | 11760 |

Reply:
I have thought about it some more and came up with yet another conversion strategy: creating custom `ColorVal` conformance to `Sequence` protocol by implementing `SwizzlingIterator`.

<details>

````swift
extension ColorVal {
  public struct SwizzlingIterator {
    let c: ColorVal
    var i: UInt8 = 0
    init(_ c: ColorVal) { self.c = c }
  }
}

extension ColorVal.SwizzlingIterator: IteratorProtocol {
  public mutating func next() -> UInt8? {
    i = i &+ 1
    switch i {
    case 1: return c.a
    case 2: return c.r
    case 3: return c.g
    case 4: return c.b
    default: return nil
    }
  }
}

extension ColorVal: Sequence {
  public func makeIterator() -> SwizzlingIterator {
    return SwizzlingIterator(self)
  }
}
````
</details>

These are the performance results for following variants:

| Variant | Min | Q1 | Med | Q3 | Max | Code |
|---|---:|---:|---:|---:|---:|---|
| flatMap.SwizzleIt | 564 | 573 | 580 | 602 | 692 | `input.flatMap { $0 }` |
| lazy.flatMap.SwizzleIt | 319 | 319 | 319 | 319 | 319 | `input.lazy.flatMap { $0 }` |
| joined.SwizzleIt | 319 | 319 | 319 | 319 | 319 | `input.joined()` |
| lazy.joined.SwizzleIt | 319 | 319 | 319 | 319 | 319 | `input.lazy.joined()` |

What do you make of this? Is this worth including? It looks like the last three did produce the exact same code after optimization...

I have also tried `flatMap` and `map.joined` using `ContiguousArray([$0.a, $0.r, $0.g, $0.b])` in the closure, expecting to see some improvement for `Color.Ref`, but saw no significant change compared to plain `Array`. Was I using it right with the array literal syntax?

Reply:
That's it. Byte-By-Byte: **136**! Thank you Andrew.

Reply:
>  If you find something faster then this then I'd be interested to know why

😏 Building up from `SwizzleIt` from above... conforming ColorVal to `RandomAccessCollection` gives runtime of **82** for `joined`, `lazy.flatMap` and `lazy.joined`

````swift
extension ColorVal: RandomAccessCollection {
  public var count: Int { return 4 }
  public var startIndex: Int { return 0 }
  public var endIndex: Int { return 4 }
  public func index(after i: Int) -> Int { return i &+ 1 }
  public subscript(position: Int) -> UInt8 {
    switch position {
    case 0: return a
    case 1: return r
    case 2: return g
    case 3: return b
    default: return 0 // never happens
    }
  }
}
````
Profiling this in instruments, the assembler code looks ordinary. So maybe it could be even faster, if it did swizzling using vector instructions?

Reply:
Hmm... 🧐🤨 I've been exploring the disassembled code of the fastest variant over at https://swift.godbolt.org/ … and now I'm not sure how much sense it makes, the way I've been comparing lazy versions with eager ones. When I've switched to unsafe addition for counting the elements in the `CountLast`, that record time 82 went down to 51. But of course it's fast to get to the last element by exhausting the sequence, when you don't have to do anything else... But on the other hand, switching to fully materializing the lazy sequences with `Array.init(Sequence)` gives unfair advantage to the eager versions and muddles the water by mixing in the Array's sequence/collection initializer with it's optimization baggage. I'll explore dumping the sequence into preallocated `UnsafeBufferPointer`.

Reply:
It looks like the improvement potential from using inlined lazy sequences and collections is real. I've replaced the `CountLast` with `materializeSequence`:

````swift
struct CountLast<T> {
  var count: Int = 0
  var last: T? = nil
  static func collect(_ cl: inout CountLast, _ element: T){
    cl.count += 1
    cl.last = element
  }
}

// Helper function to fully materialize Sequence into preallocated buffer.
// Cheats by knowing `flatSize` in advance. But demonstrates the power of
// inlining the Sequence initializer.
func materializeSequence<S: Sequence>(_ s: S) -> [S.Element]
  where S.Element: Initializable {
  var a = ContiguousArray<S.Element>(repeating: S.Element(), count: flatSize)
  a.withUnsafeMutableBufferPointer { buffer in
    var i = 0
    _ = s.reduce(into:buffer) {
      $0[i] = $1
      i &+= 1
    }
  }
  return Array(a)
}
````


<details open>
  <summary><strong>Improvement (6)</strong></summary>
  
TEST                                     | OLD   | NEW   | DELTA  | RATIO     
---                                      | ---   | ---   | ---    | ---       
Flatten.ColorVal.lazy.map.joined.SwizSeq | 319   | 234   | -26.6% | **1.36x** 
Flatten.ColorVal.lazy.flatMap.SwizSeq    | 319   | 234   | -26.6% | **1.36x** 
Flatten.ColorVal.lazy.flatMap.SwizCol    | 82    | 63    | -23.2% | **1.30x** 
Flatten.ColorVal.lazy.map.joined.SwizCol | 82    | 63    | -23.2% | **1.30x** 
Flatten.ColorVal.map.joined.SwizSeq      | 540   | 488   | -9.6%  | **1.11x** 
Flatten.ColorVal.map.joined.SwizCol      | 291   | 277   | -4.8%  | **1.05x (?)** 

</details>

<details>
  <summary>Regression (8)</summary>
  
TEST                                     | OLD   | NEW   | DELTA  | RATIO     
---                                      | ---   | ---   | ---    | ---       
Flatten.AnySeq.LazySeq.Tuple4.flatMap    | 1001  | 1172  | +17.1% | **0.85x** 
Flatten.LazySeq.Tuple4.for-in            | 942   | 1062  | +12.7% | **0.89x** 
Flatten.AnySeq.LazySeq.Tuple4.for-in     | 975   | 1083  | +11.1% | **0.90x** 
Flatten.LazySeq.Array4.for-in            | 8103  | 8725  | +7.7%  | **0.93x (?)** 
Flatten.ColorVal.map.joined.Array        | 10575 | 11258 | +6.5%  | **0.94x (?)** 
Flatten.Array.Array4.lazy.flatMap        | 1297  | 1377  | +6.2%  | **0.94x** 
Flatten.Array.Array4.lazy.joined         | 1298  | 1376  | +6.0%  | **0.94x** 

</details>
<details >
  <summary>No Changes (26)</summary>
  
TEST                                     | OLD   | NEW   | DELTA  | RATIO     
---                                      | ---   | ---   | ---    | ---       
Flatten.AnySeq.LazySeq.Tuple4.map.joined | 11280 | 11469 | +1.7%  | 0.98x (?) 
Flatten.Array.Array4.joined              | 309   | 309   | +0.0%  | 1.00x     
Flatten.Array.Tuple4.lazy.flatMap        | 7058  | 7203  | +2.1%  | 0.98x (?) 
Flatten.Array.Tuple4.lazy.map.joined     | 7048  | 7199  | +2.1%  | 0.98x (?) 
Flatten.Array.Tuple4.map.joined          | 10094 | 10090 | -0.0%  | 1.00x (?) 
Flatten.ColorRef.lazy.flatMap.Array      | 8040  | 8042  | +0.0%  | 1.00x (?) 
Flatten.ColorRef.lazy.flatMap.ContArr    | 7908  | 7998  | +1.1%  | 0.99x (?) 
Flatten.ColorRef.lazy.flatMap.SwizCol    | 1072  | 1123  | +4.8%  | 0.95x     
Flatten.ColorRef.lazy.flatMap.SwizSeq    | 1242  | 1278  | +2.9%  | 0.97x     
Flatten.ColorRef.lazy.map.joined.Array   | 7996  | 8072  | +1.0%  | 0.99x (?) 
Flatten.ColorRef.lazy.map.joined.ContArr | 7852  | 8064  | +2.7%  | 0.97x (?) 
Flatten.ColorRef.lazy.map.joined.SwizCol | 1072  | 1117  | +4.2%  | 0.96x     
Flatten.ColorRef.lazy.map.joined.SwizSeq | 1246  | 1281  | +2.8%  | 0.97x     
Flatten.ColorRef.map.joined.Array        | 11859 | 11829 | -0.3%  | 1.00x (?) 
Flatten.ColorRef.map.joined.ContArr      | 11437 | 11525 | +0.8%  | 0.99x (?) 
Flatten.ColorRef.map.joined.SwizCol      | 2347  | 2395  | +2.0%  | 0.98x (?) 
Flatten.ColorRef.map.joined.SwizSeq      | 2538  | 2512  | -1.0%  | 1.01x (?) 
Flatten.ColorVal.lazy.flatMap.Array      | 7056  | 7074  | +0.3%  | 1.00x (?) 
Flatten.ColorVal.lazy.flatMap.ContArr    | 7092  | 6981  | -1.6%  | 1.02x (?) 
Flatten.ColorVal.lazy.map.joined.Array   | 7094  | 7035  | -0.8%  | 1.01x (?) 
Flatten.ColorVal.lazy.map.joined.ContArr | 7050  | 7034  | -0.2%  | 1.00x (?) 
Flatten.ColorVal.map.joined.ContArr      | 10313 | 10391 | +0.8%  | 0.99x (?) 
Flatten.LazySeq.Array4.flatMap           | 7304  | 7601  | +4.1%  | 0.96x (?) 
Flatten.LazySeq.Array4.joined            | 7308  | 7570  | +3.6%  | 0.97x (?) 
Flatten.LazySeq.Tuple4.flatMap           | 7298  | 7417  | +1.6%  | 0.98x (?) 
Flatten.LazySeq.Tuple4.map.joined        | 7298  | 7516  | +3.0%  | 0.97x (?) 

</details>

These regressions don't concern me much, because it looks like using `Array` in the closures is generally some kind of an optimization barrier...

I have initially experimented with materializing the sequences using Array's initializer (`let fa = Array(f)`), but it was never faster than the imperative approach. Which surprised me, because`.lazy` is the performance magic, [in my experience](https://forums.swift.org/t/troubling-sequence/13130)... The old comment next to template using `CountLast` justified its necessity for good performance, by speculating that the measured difference was due to memory allocation overhead. But `materializeSequence` disproves this theory... So this is a bit confusing now. 

Array's sequence initializer seems to work well for `Tuple4.Unsafe.InitSeq` and also for `ColorVal` variants `Unsafe.UInt32InitSeq` and `Unsafe.ColorValInitSeq`. But the likes of `ColorVal.lazy.flatMap.SwizCol` are not getting the boost I'd expect when plugged into `Array.init<Sequence>`. They are clearly best with `CountLast` and also with `materializeSequence`, even better than `Unsafe.BytesReserve` (139 μs)!

It looks like some performance bug to me… @atrick how would you handle this? Do we need to fix this before ABI freeze? 
// cc @airspeedswift

#### `materializeSequence(f)` vs. `Array(f)`
<details>
  <summary>Regression (17)</summary>
  
TEST                                     | OLD   | NEW   | DELTA    | RATIO     
---                                      | ---   | ---   | ---      | ---       
Flatten.ColorVal.lazy.map.joined.SwizCol | 63    | 716   | +1036.5% | **0.09x** 
Flatten.ColorVal.lazy.flatMap.SwizCol    | 63    | 675   | +971.4%  | **0.09x** 
Flatten.ColorVal.lazy.map.joined.SwizSeq | 234   | 891   | +280.8%  | **0.26x** 
Flatten.ColorVal.lazy.flatMap.SwizSeq    | 234   | 850   | +263.2%  | **0.28x** 
Flatten.ColorVal.map.joined.SwizCol      | 277   | 677   | +144.4%  | **0.41x** 
Flatten.ColorVal.map.joined.SwizSeq      | 488   | 899   | +84.2%   | **0.54x** 
Flatten.LazySeq.Array4.flatMap           | 7601  | 9100  | +19.7%   | **0.84x (?)** 
Flatten.ColorRef.lazy.map.joined.SwizCol | 1117  | 1326  | +18.7%   | **0.84x** 
Flatten.ColorRef.lazy.flatMap.SwizCol    | 1123  | 1325  | +18.0%   | **0.85x** 
Flatten.Array.Array4.lazy.flatMap        | 1377  | 1624  | +17.9%   | **0.85x** 
Flatten.Array.Array4.lazy.joined         | 1376  | 1608  | +16.9%   | **0.86x** 
Flatten.ColorRef.map.joined.SwizSeq      | 2512  | 2778  | +10.6%   | **0.90x** 
Flatten.LazySeq.Tuple4.flatMap           | 7417  | 8189  | +10.4%   | **0.91x (?)** 
Flatten.LazySeq.Tuple4.map.joined        | 7516  | 8252  | +9.8%    | **0.91x** 
Flatten.ColorRef.lazy.flatMap.SwizSeq    | 1278  | 1384  | +8.3%    | **0.92x** 
Flatten.ColorRef.lazy.map.joined.SwizSeq | 1281  | 1383  | +8.0%    | **0.93x** 
Flatten.ColorRef.map.joined.SwizCol      | 2395  | 2551  | +6.5%    | **0.94x** 

</details>

<details >
  <summary>Improvement (4)</summary>
  
TEST                                     | OLD   | NEW   | DELTA    | RATIO     
---                                      | ---   | ---   | ---      | ---       
Flatten.AnySeq.LazySeq.Tuple4.flatMap    | 1172  | 1005  | -14.2%   | **1.17x** 
Flatten.LazySeq.Tuple4.for-in            | 1062  | 938   | -11.7%   | **1.13x** 
Flatten.AnySeq.LazySeq.Tuple4.for-in     | 1083  | 969   | -10.5%   | **1.12x** 
Flatten.LazySeq.Array4.for-in            | 8725  | 8234  | -5.6%    | **1.06x (?)** 

</details>

<details >
  <summary>No Changes (18)</summary>
  
TEST                                     | OLD   | NEW   | DELTA    | RATIO     
---                                      | ---   | ---   | ---      | ---       
Flatten.AnySeq.LazySeq.Tuple4.map.joined | 11469 | 11691 | +1.9%    | 0.98x (?) 
Flatten.Array.Array4.joined              | 309   | 309   | +0.0%    | 1.00x     
Flatten.Array.Tuple4.lazy.flatMap        | 7203  | 7455  | +3.5%    | 0.97x (?) 
Flatten.Array.Tuple4.lazy.map.joined     | 7199  | 7511  | +4.3%    | 0.96x (?) 
Flatten.Array.Tuple4.map.joined          | 10090 | 10108 | +0.2%    | 1.00x (?) 
Flatten.ColorRef.lazy.flatMap.Array      | 8042  | 8208  | +2.1%    | 0.98x (?) 
Flatten.ColorRef.lazy.flatMap.ContArr    | 7998  | 8175  | +2.2%    | 0.98x (?) 
Flatten.ColorRef.lazy.map.joined.Array   | 8072  | 8218  | +1.8%    | 0.98x (?) 
Flatten.ColorRef.lazy.map.joined.ContArr | 8064  | 8216  | +1.9%    | 0.98x (?) 
Flatten.ColorRef.map.joined.Array        | 11829 | 12089 | +2.2%    | 0.98x (?) 
Flatten.ColorRef.map.joined.ContArr      | 11525 | 11837 | +2.7%    | 0.97x (?) 
Flatten.ColorVal.lazy.flatMap.Array      | 7074  | 7247  | +2.4%    | 0.98x (?) 
Flatten.ColorVal.lazy.flatMap.ContArr    | 6981  | 7172  | +2.7%    | 0.97x (?) 
Flatten.ColorVal.lazy.map.joined.Array   | 7035  | 7226  | +2.7%    | 0.97x (?) 
Flatten.ColorVal.lazy.map.joined.ContArr | 7034  | 7225  | +2.7%    | 0.97x (?) 
Flatten.ColorVal.map.joined.Array        | 11258 | 10976 | -2.5%    | 1.03x (?) 
Flatten.ColorVal.map.joined.ContArr      | 10391 | 10747 | +3.4%    | 0.97x (?) 
Flatten.LazySeq.Array4.joined            | 7570  | 7664  | +1.2%    | 0.99x (?) 

</details>

Reply:
> It looks like some performance bug to me… @atrick how would you handle this? Do we need to fix this before ABI freeze?
// cc @airspeedswift

We can replace `@inlinable` stdlib implementations with semantically equivalent ones without breaking ABI. We can also theoretically make things `@inlinable` that weren't before. Without debugging it further, I think this is just a normal bug against the optimizer, or rather an opportunity for better performance.